File: android/app/src/main/java/com/nearpair/app/NearPairViewModel.kt
package com.nearpair.app

import android.app.Application
import android.net.Uri
import android.os.Build
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import com.nearpair.app.files.FileStore
import com.nearpair.app.model.FailureReason
import com.nearpair.app.model.ReceivedFile
import com.nearpair.app.model.StagedFile
import com.nearpair.app.model.TransferState
import com.nearpair.app.nearby.NearbyTransferEngine
import com.nearpair.app.nearby.TransferEngine
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import java.io.File

class NearPairViewModel(application: Application) : AndroidViewModel(application) {
    private val fileStore = FileStore(application)
    private val engine: TransferEngine = NearbyTransferEngine(application, fileStore)

    val state: StateFlow<TransferState> = engine.state
    val devices = engine.devices
    val receivedFile: StateFlow<ReceivedFile?> = engine.receivedFile

    private val _isStaging = MutableStateFlow(false)
    val isStaging: StateFlow<Boolean> = _isStaging.asStateFlow()

    private val _uiError = MutableStateFlow<String?>(null)
    val uiError: StateFlow<String?> = _uiError.asStateFlow()

    private val _uiErrorRecovery = MutableStateFlow(UiErrorRecovery.NONE)
    val uiErrorRecovery: StateFlow<UiErrorRecovery> = _uiErrorRecovery.asStateFlow()

    private val _shareRequests = MutableStateFlow<ShareRequest?>(null)
    val shareRequests: StateFlow<ShareRequest?> = _shareRequests.asStateFlow()

    private var pendingOutgoing: StagedFile? = null
    private var sendStartedForEndpoint: String? = null
    private var lastRole: Role? = null
    private var appInForeground = false

    private val _deviceName = MutableStateFlow(
        listOf(Build.MANUFACTURER, Build.MODEL)
            .filter { it.isNotBlank() }
            .joinToString(" ")
            .take(48)
            .ifBlank { "Android device" },
    )
    val deviceName: StateFlow<String> = _deviceName.asStateFlow()

    init {
        viewModelScope.launch {
            state.collect { current ->
                if (current is TransferState.Connected && pendingOutgoing != null && sendStartedForEndpoint != current.endpointId) {
                    sendStartedForEndpoint = current.endpointId
                    engine.sendFile(requireNotNull(pendingOutgoing))
                }
            }
        }
    }

    fun stageForDirectTransfer(uri: Uri) {
        lastRole = Role.SEND
        _uiError.value = null
        _uiErrorRecovery.value = UiErrorRecovery.NONE
        _isStaging.value = true
        viewModelScope.launch {
            fileStore.stageOutgoing(uri)
                .onSuccess { staged ->
                    if (appInForeground) {
                        pendingOutgoing?.let(fileStore::delete)
                        pendingOutgoing = staged
                        sendStartedForEndpoint = null
                        engine.discoverNearbyDevices(deviceName.value)
                    } else {
                        fileStore.delete(staged)
                        _uiError.value = "NearPair must remain open while preparing a direct transfer. Return and tap Send again."
                    }
                }
                .onFailure { _uiError.value = it.localizedMessage ?: "The selected file could not be staged." }
            _isStaging.value = false
        }
    }

    fun stageForSystemShare(uri: Uri) {
        _uiError.value = null
        _uiErrorRecovery.value = UiErrorRecovery.NONE
        _isStaging.value = true
        viewModelScope.launch {
            fileStore.stageOutgoing(uri)
                .onSuccess { staged ->
                    _shareRequests.value = ShareRequest(
                        file = staged.file,
                        displayName = staged.displayName,
                        mimeType = staged.mimeType,
                    )
                }
                .onFailure { _uiError.value = it.localizedMessage ?: "The selected file could not be shared." }
            _isStaging.value = false
        }
    }

    fun startReceiving() {
        lastRole = Role.RECEIVE
        _uiError.value = null
        engine.startReceiving(deviceName.value)
    }

    fun requestConnection(endpointId: String) {
        val device = devices.value.firstOrNull { it.endpointId == endpointId } ?: return
        engine.requestConnection(device, deviceName.value)
    }

    fun acceptConnection(endpointId: String) = engine.acceptConnection(endpointId)

    fun rejectConnection(endpointId: String) = engine.rejectConnection(endpointId)

    fun cancel() = engine.cancelTransfer()

    fun retry() {
        _uiError.value = null
        sendStartedForEndpoint = null
        engine.reset()
        when (lastRole) {
            Role.SEND -> if (pendingOutgoing != null) engine.discoverNearbyDevices(deviceName.value)
            Role.RECEIVE -> engine.startReceiving(deviceName.value)
            null -> Unit
        }
    }

    fun done() {
        pendingOutgoing?.let(fileStore::delete)
        pendingOutgoing = null
        sendStartedForEndpoint = null
        lastRole = null
        engine.reset()
        _uiError.value = null
        _uiErrorRecovery.value = UiErrorRecovery.NONE
    }

    fun permissionDenied() {
        _uiErrorRecovery.value = UiErrorRecovery.APP_PERMISSIONS
        _uiError.value = "Allow NearPair's Nearby devices permission on the app settings screen, then return and retry. Bluetooth and Wi-Fi are turned on separately in Android's Connections settings."
    }

    fun radiosDisabled(radios: List<String>) {
        _uiErrorRecovery.value = UiErrorRecovery.CONNECTIONS_SETTINGS
        _uiError.value = "Turn on ${radios.joinToString(" and ")} in Android's Connections settings, then return and retry."
    }

    fun dismissUiError() {
        _uiError.value = null
        _uiErrorRecovery.value = UiErrorRecovery.NONE
    }

    fun updateDeviceName(value: String) {
        _deviceName.value = value.take(48)
    }

    fun shareReceived() {
        val received = receivedFile.value ?: return
        _shareRequests.value = ShareRequest(received.file, received.displayName, received.mimeType)
    }

    fun shareRequestHandled(request: ShareRequest) {
        // Android has no reliable callback for when the selected share target
        // finishes reading a granted URI. Keep this cache file available and
        // let FileStore's stale-file cleanup remove it on a later app start.
        _shareRequests.compareAndSet(request, null)
    }

    fun saveReceived(destination: Uri) {
        val received = receivedFile.value ?: return
        viewModelScope.launch {
            fileStore.copyReceivedTo(received, destination)
                .onSuccess {
                    fileStore.delete(received)
                    done()
                }
                .onFailure { _uiError.value = it.localizedMessage ?: "The file could not be saved." }
        }
    }

    fun deleteReceived() {
        receivedFile.value?.let(fileStore::delete)
        done()
    }

    fun onAppForegrounded() {
        appInForeground = true
    }

    fun onAppBackgrounded() {
        appInForeground = false
        engine.onAppBackgrounded()
    }

    override fun onCleared() {
        pendingOutgoing?.let(fileStore::delete)
        engine.close()
        super.onCleared()
    }

    data class ShareRequest(
        val file: File,
        val displayName: String,
        val mimeType: String,
    )

    enum class UiErrorRecovery { NONE, APP_PERMISSIONS, CONNECTIONS_SETTINGS }

    private enum class Role { SEND, RECEIVE }
}

File: ios/NearPair/NearPairViewModel.swift
import Combine
import Foundation
import PhotosUI
import UIKit

@MainActor
final class NearPairViewModel: ObservableObject {
    @Published private(set) var phase: TransferPhase = .idle
    @Published private(set) var devices: [NearbyDevice] = []
    @Published private(set) var receivedFile: ReceivedFile?
    @Published private(set) var isStaging = false
    @Published var alertMessage: String?
    @Published var shareItem: ShareItem?
    @Published var exportItem: ExportItem?
    @Published var deviceName = String(UIDevice.current.name.prefix(48))

    private let fileStore: FileStore
    private let engine: NearbyTransferEngine
    private var subscriptions: Set<AnyCancellable> = []
    private var pendingOutgoing: StagedFile?
    private var sendStartedForEndpoint: String?
    private var lastRole: Role?

    init() {
        let store = FileStore()
        fileStore = store
        engine = NearbyTransferEngine(fileStore: store)

        engine.$phase
            .sink { [weak self] phase in
                guard let self else { return }
                self.phase = phase
                if case .connected(let endpointID, _) = phase,
                   let file = self.pendingOutgoing,
                   self.sendStartedForEndpoint != endpointID {
                    self.sendStartedForEndpoint = endpointID
                    self.engine.sendFile(file)
                }
            }
            .store(in: &subscriptions)

        engine.$devices.assign(to: &$devices)
        engine.$receivedFile.assign(to: &$receivedFile)
    }

    func chooseDirectFile(_ url: URL) {
        lastRole = .send
        stage(url: url, purpose: .direct)
    }

    func chooseSystemShareFile(_ url: URL) {
        stage(url: url, purpose: .systemShare)
    }

    func chooseMedia(_ item: PhotosPickerItem) {
        lastRole = .send
        alertMessage = nil
        isStaging = true
        Task {
            defer { isStaging = false }
            do {
                guard let media = try await item.loadTransferable(type: PickedMedia.self) else {
                    throw ViewModelError.photoUnavailable
                }
                let staged = try await fileStore.stageOwnedTemporary(media.url)
                pendingOutgoing = staged
                sendStartedForEndpoint = nil
                engine.discoverNearbyDevices(deviceName: deviceName)
            } catch {
                alertMessage = error.localizedDescription
            }
        }
    }

    func startReceiving() {
        lastRole = .receive
        alertMessage = nil
        engine.startReceiving(deviceName: deviceName)
    }

    func requestConnection(_ device: NearbyDevice) {
        engine.requestConnection(to: device, localDeviceName: deviceName)
    }

    func accept(endpointID: String) { engine.acceptConnection(endpointID: endpointID) }
    func reject(endpointID: String) { engine.rejectConnection(endpointID: endpointID) }
    func cancel() { engine.cancelTransfer() }

    func retry() {
        alertMessage = nil
        sendStartedForEndpoint = nil
        engine.reset()
        switch lastRole {
        case .send where pendingOutgoing != nil:
            engine.discoverNearbyDevices(deviceName: deviceName)
        case .receive:
            engine.startReceiving(deviceName: deviceName)
        default:
            break
        }
    }

    func done() {
        let outgoingURL = pendingOutgoing?.url
        pendingOutgoing = nil
        sendStartedForEndpoint = nil
        lastRole = nil
        engine.reset()
        if let outgoingURL { Task { await fileStore.delete(url: outgoingURL) } }
    }

    func shareReceived() {
        guard let receivedFile else { return }
        shareItem = ShareItem(url: receivedFile.url, cleanupAfterShare: false)
    }

    func exportReceived() {
        guard let receivedFile else { return }
        exportItem = ExportItem(url: receivedFile.url)
    }

    func exportFinished(saved: Bool) {
        guard let item = exportItem else { return }
        exportItem = nil
        if saved {
            Task {
                await fileStore.delete(url: item.url)
                done()
            }
        }
    }

    func shareFinished() {
        guard let item = shareItem else { return }
        shareItem = nil
        if item.cleanupAfterShare {
            Task { await fileStore.delete(url: item.url) }
        }
    }

    func deleteReceived() {
        guard let receivedFile else { return }
        Task {
            await fileStore.delete(url: receivedFile.url)
            done()
        }
    }

    func onAppBackgrounded() { engine.onAppBackgrounded() }

    private func stage(url: URL, purpose: StagePurpose) {
        alertMessage = nil
        isStaging = true
        Task {
            defer { isStaging = false }
            do {
                let staged = try await fileStore.stageOutgoing(url)
                switch purpose {
                case .direct:
                    pendingOutgoing = staged
                    sendStartedForEndpoint = nil
                    engine.discoverNearbyDevices(deviceName: deviceName)
                case .systemShare:
                    shareItem = ShareItem(url: staged.url, cleanupAfterShare: true)
                }
            } catch {
                alertMessage = error.localizedDescription
            }
        }
    }

    struct ShareItem: Identifiable {
        let id = UUID()
        let url: URL
        let cleanupAfterShare: Bool
    }

    struct ExportItem: Identifiable {
        let id = UUID()
        let url: URL
    }

    private enum Role { case send, receive }
    private enum StagePurpose { case direct, systemShare }
}

private enum ViewModelError: LocalizedError {
    case photoUnavailable

    var errorDescription: String? { "The selected photo or video could not be loaded." }
}
